### ISSUE TITLE
"User not found" error is thrown while the user is valid and in the database

### ISSUE THREAD

**Issue by cgm-maarten** [keycloak/keycloak#34363]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

infinispan

### Describe the bug

User enters a correct username. (In the username form)
The user gets a "Invalid username" error. (in the log there is a user_not_found message for the username)

The user can be found with this username in the admin interface, and when I try it myself with that username (copy pasted form the admin user page.) I get the same error.

When I change the username (by adding a character) save the account and change it back, the problem disappears, the user can be normally found again. The username consists of normal characters (a-z)

It has happened to multiple users (but not all)

### Version

26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

After upgrading everyone can still use their username to login and find users

### Actual behavior

not all users can be found.

### How to Reproduce?

Unfortunately I do not have any idea how to reproduce it.

### Anything else?

We have our installation migrated from 25.0.4 to 26.0.0.
We didn't use persistent sessions in the 25.0.4. (And I believe it was turned on by default in 26.0.0) (might be relevant)

I suspect it might have something to do with the user cache, since updating the username fixes the problem for a specific user, but I can't do that for all users, to fix the problem for everyone.

Log message:

2024-10-25 21:33:18,935 WARN  [org.keycloak.events] (executor-thread-35) type="LOGIN_ERROR", realmId="xxx", realmName="xxx", clientId="xx", userId="null", ipAddress="xxx", error="user_not_found", auth_method="openid-connect", auth_type="code", redirect_uri="valid uri", code_id="xxx", username="<anonimized, regular a-z characters>"

**Comment by ahus1** [comment]
Thank you for reporting this. Some additional questions to analyze this: 

* Do your users login via username or via email or both? Have the users that are not found been modified recently? 
* Does it fail when they log in via the username, or does it fail when logging in with the email address? 
* Where do you store your user - in a database? in LDAP?

cc: @pruivo - any ideas or additional questions?

**Comment by pruivo** [comment]
@ahus1 I don't have any idea. It looks like a database issue since we don't cache database misses in the `users` cache. AFAIK, `user_not_found` means the user was not found in the cache and the database. The `organization` feature is enabled by default. Does it affect the user's table in any way?

**Comment by cgm-maarten** [comment]
We didn't configure organizations.

**Comment by pedroigor** [comment]
I haven't seen this behavior so far and I'm not sure how to move it forward. We need some way to reproduce this but perhaps can you tell us if you have the same behavior from any older version?

**Comment by cgm-maarten** [comment]
The problem is that I cannot reproduce it (I have found a work around, which resolved the problem)
We only noticed the problem on production, so we cannot roll back and try again with different versions.

But I think the query that is used for retrieving users using the username-form uses a different method of comparing usernames than the admin panel search. I think that could be a lead.

**Comment by pedroigor** [comment]
The Admin API basically uses a like search, whether it is an exact match during the login. They are different queries.

I'm not sure what you mean by "ignores hidden characters that are caught by the wildcard" but I don't see how caching is related to that. What you describe is hard to follow without reproducing it and see what is going on.

I would say that if we had issues when querying and caching users we would have a lot of thumbs up here.

For this reason, I think we can close this issue for now and keep an eye on similar reports. Or better steps on how to reproduce the problem. 

I personally don't link these issues as they might indicate a problem. But hard to tell ...

**Comment by cgm-maarten** [comment]
I agree, let's close the issue. 

I don't know how it could be related, other than the workaround: find the user, save it again, and the user can log in again: so there I thought: might be some trimming going on at save. So exact search doesn't find the username, but finding in admin does retrieve result. That's what's weirding me out.

And I might be wrong on the caching: restarting the application didn't resolve the issue. (But then as of 26.0.0 we have persistent sessions)

**Comment by renannprado** [comment]
For whoever is having this issue, I've faced something similar with user not found, although might be completely unrelated to this.

In my case I'm using bitwarden as a password manager and I use it to login in the admin console of keycloak by using the chrome bitwarden extension.

Turns out that suddenly* login started to fail, I even ended up creating a new admin user via command line because I was locked out.

Looking into the admin dashboard with the new user, I see that the user is there and password hasn't changed, but I still cannot login.

Then I went to chrome's network tab I noticed that my email was prepended with some invisible characters (???), looking at the encoded version of the email this is that I got:

```
%E2%81%A0myadminuser@example.com
```

If you URL _decode_ the above string, you'll see that those are all invisible characters before the `myadminuser...` starts.

When I tried to type the email myself (instead of filling it out with bitwarden), I could login.

Then I updated the entry in bitwarden by writing the email myself and everything is fine again.

*IDK if due to a bug in bitwarden/extension or someone messed up with the entry in bitwarden itself
